pkg/solana/logpoller/parser.go

```diff
@@ -0,0 +1,194 @@
+package logpoller
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type MessageStyle string
+
+const (
+	MessageStyleMuted   MessageStyle = "muted"
+	MessageStyleInfo    MessageStyle = "info"
+	MessageStyleSuccess MessageStyle = "success"
+	MessageStyleWarning MessageStyle = "warning"
+)
+
+var (
+	invokeMatcher      = regexp.MustCompile(`Program (\w*) invoke \[(\d)\]`)
+	consumedMatcher    = regexp.MustCompile(`Program \w* consumed (\d*) (.*)`)
+	logMatcher         = regexp.MustCompile(`Program log: (.*)`)
+	dataMatcher        = regexp.MustCompile(`Program data: (.*)`)
+	instructionMatcher = regexp.MustCompile(`Instruction: (.*)`)
+)
+
+type BlockData struct {
+	BlockNumber         uint64
+	BlockHash           string
+	TransactionHash     string
+	TransactionIndex    int
+	TransactionLogIndex uint
+}
+
+type ProgramLog struct {
+	BlockData
+	Text   string
+	Prefix string
+	Style  MessageStyle
+}
+
+type ProgramEvent struct {
+	BlockData
+	Prefix       string
+	FunctionName string
+	Data         string
+}
+
+type ProgramOutput struct {
+	Program      string
+	Logs         []ProgramLog
+	Events       []ProgramEvent
+	ComputeUnits uint
+	Truncated    bool
+	Failed       bool
+}
+
+func prefixBuilder(depth int) string {
+	return strings.Repeat(">", depth)
+}
+
+/*
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 invoke [1]
+Program log: Instruction: CreateLog
+Program data: HDQnaQjSWwkNAAAASGVsbG8sIFdvcmxkISoAAAAAAAAA // base64 encoded; borsh encoded with identifier
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 consumed 1477 of 200000 compute units
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 success
+*/
+func parseProgramLogs(logs []string) []ProgramOutput {
```

Comment:
Should we have test coverage for this func? It seems to be very complex and can be broken easily.

Reply:
Some real-world solana logs would be very helpful for this. I am simply basing the parser off the logs that I was able to create with the test contract and the research I did.

Reply:
Can't we have a test that uses logs that you've found? Just to ensure we lock implementation and it's more difficult to break something that was previously working.